File: Projetos/ACBrLib/Demos/C#/Boleto/Imports/Dinamico/MT/ACBrBoleto.cs
using System;
using System.IO;
using System.Net.Security;
using System.Text;
using ACBrLib.Core;
using ACBrLib.Core.Boleto;

namespace ACBrLib.Boleto
{
    public sealed partial class ACBrBoleto : ACBrLibHandle
    {
        #region Constructors

        public ACBrBoleto(string eArqConfig = "", string eChaveCrypt = "") : base(IsWindows ? "ACBrBoleto64.dll" : "libacbrboleto64.so",
                                                                                  IsWindows ? "ACBrBoleto32.dll" : "libacbrboleto32.so")
        {
            var inicializar = GetMethod<Boleto_Inicializar>();
            var ret = ExecuteMethod(() => inicializar(ref libHandle, ToUTF8(eArqConfig), ToUTF8(eChaveCrypt)));

            CheckResult(ret);

            Config = new ACBrBoletoConfig(this);
        }

        #endregion Constructors

        #region Properties

        public string Nome
        {
            get
            {
                var bufferLen = BUFFER_LEN;
                var buffer = new StringBuilder(bufferLen);

                var method = GetMethod<Boleto_Nome>();
                var ret = ExecuteMethod(() => method(libHandle, buffer, ref bufferLen));

                CheckResult(ret);

                return ProcessResult(buffer, bufferLen);
            }
        }

        public string Versao
        {
            get
            {
                var bufferLen = BUFFER_LEN;
                var buffer = new StringBuilder(bufferLen);

                var method = GetMethod<Boleto_Versao>();
                var ret = ExecuteMethod(() => method(libHandle, buffer, ref bufferLen));

                CheckResult(ret);

                return ProcessResult(buffer, bufferLen);
            }
        }

        public ACBrBoletoConfig Config { get; }

        #endregion Properties

        #region Methods

        #region Ini

        public override void ConfigGravar(string eArqConfig = "")
        {
            var gravarIni = GetMethod<Boleto_ConfigGravar>();
            var ret = ExecuteMethod(() => gravarIni(libHandle, ToUTF8(eArqConfig)));

            CheckResult(ret);
        }

        public override void ConfigLer(string eArqConfig = "")
        {
            var lerIni = GetMethod<Boleto_ConfigLer>();
            var ret = ExecuteMethod(() => lerIni(libHandle, ToUTF8(eArqConfig)));

            CheckResult(ret);
        }

        public override T ConfigLerValor<T>(ACBrSessao eSessao, string eChave)
        {
            var method = GetMethod<Boleto_ConfigLerValor>();

            var bufferLen = BUFFER_LEN;
            var pValue = new StringBuilder(bufferLen);
            var ret = ExecuteMethod(() => method(libHandle, ToUTF8(eSessao.ToString()), ToUTF8(eChave), pValue, ref bufferLen));
            CheckResult(ret);

            var value = ProcessResult(pValue, bufferLen);
            return ConvertValue<T>(value);
        }

        public override void ConfigGravarValor(ACBrSessao eSessao, string eChave, object value)
        {
            if (value == null) return;

            var method = GetMethod<Boleto_ConfigGravarValor>();
            var propValue = ConvertValue(value);

            var ret = ExecuteMethod(() => method(libHandle, ToUTF8(eSessao.ToString()), ToUTF8(eChave), ToUTF8(propValue)));
            CheckResult(ret);
        }

        public override void ImportarConfig(string eArqConfig = "")
        {
            var importarConfig = GetMethod<Boleto_ConfigImportar>();
            var ret = ExecuteMethod(() => importarConfig(libHandle, ToUTF8(eArqConfig)));

            CheckResult(ret);
        }

        public override string ExportarConfig()
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Boleto_ConfigExportar>();
            var ret = ExecuteMethod(() => method(libHandle, buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen);
        }

        #endregion Ini

        public void ConfigurarDados(params BoletoInfo[] infos)
        {
            var iniFile = new ACBrIniFile();
            foreach (var info in infos)
                info.WriteToIni(iniFile);

            ConfigurarDados(iniFile.ToString());
        }

        public void ConfigurarDados(string eArquivoIni)
        {
            var method = GetMethod<Boleto_ConfigurarDados>();
            var ret = ExecuteMethod(() => method(libHandle, ToUTF8(eArquivoIni)));

            CheckResult(ret);
        }

        public void IncluirTitulos(params Titulo[] titulos)
        {
            var iniFile = new ACBrIniFile();
            for (var i = 0; i < titulos.Length; i++)
            {
                titulos[i].Index = i + 1;
                titulos[i].WriteToIni(iniFile);
            }

            IncluirTitulos(iniFile.ToString());
        }

        public void IncluirTitulos(BoletoTpSaida eTpSaida, params Titulo[] titulos)
        {
            var iniFile = new ACBrIniFile();
            for (var i = 0; i < titulos.Length; i++)
            {
                titulos[i].Index = i + 1;
                titulos[i].WriteToIni(iniFile);
            }

            IncluirTitulos(iniFile.ToString(), eTpSaida);
        }

        public void IncluirTitulos(string eArquivoIni, BoletoTpSaida? eTpSaida = null)
        {
            var tpSaida = $"{(eTpSaida.HasValue ? (char)eTpSaida.Value : ' ')}";

            var method = GetMethod<Boleto_IncluirTitulos>();
            var ret = ExecuteMethod(() => method(libHandle, ToUTF8(eArquivoIni), ToUTF8(tpSaida)));

            CheckResult(ret);
        }

        public void LimparLista()
        {
            var method = GetMethod<Boleto_LimparLista>();
            var ret = ExecuteMethod(() => method(libHandle));
            CheckResult(ret);
        }

        public int TotalTitulosLista()
        {
            var method = GetMethod<Boleto_TotalTitulosLista>();
            var ret = ExecuteMethod(() => method(libHandle));

            CheckResult(ret);

            return ret;
        }

        public void Imprimir(string eNomeImpressora = "")
        {
            var method = GetMethod<Boleto_Imprimir>();
            var ret = ExecuteMethod(() => method(libHandle, ToUTF8(eNomeImpressora)));
            CheckResult(ret);
        }

        public void Imprimir(int indice, string eNomeImpressora = "")
        {
            var method = GetMethod<Boleto_ImprimirBoleto>();
            var ret = ExecuteMethod(() => method(libHandle, indice, ToUTF8(eNomeImpressora)));
            CheckResult(ret);
        }

        public void GerarPDF()
        {
            var method = GetMethod<Boleto_GerarPDF>();
            var ret = ExecuteMethod(() => method(libHandle));
            CheckResult(ret);
        }

        public void GerarPDF(Stream aStream)
        {
            if (aStream == null) throw new ArgumentNullException(nameof(aStream));

            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Boleto_SalvarPDF>();
            var ret = ExecuteMethod(() => method(libHandle, buffer, ref bufferLen));

            CheckResult(ret);

            var pdf = ProcessResult(buffer, bufferLen);
            Base64ToStream(pdf, aStream);
        }

        public void GerarPDF(int indice)
        {
            var method = GetMethod<Boleto_GerarPDFBoleto>();
            var ret = ExecuteMethod(() => method(libHandle, indice));
            CheckResult(ret);
        }

        public void GerarPDF(int indice, Stream aStream)
        {
            if (aStream == null) throw new ArgumentNullException(nameof(aStream));

            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Boleto_SalvarPDFBoleto>();
            var ret = ExecuteMethod(() => method(libHandle, indice, buffer, ref bufferLen));

            CheckResult(ret);

            var pdf = ProcessResult(buffer, bufferLen);
            Base64ToStream(pdf, aStream);
        }

        public void GerarHTML()
        {
            var method = GetMethod<Boleto_GerarHTML>();
            var ret = ExecuteMethod(() => method(libHandle));
            CheckResult(ret);
        }

        public void GerarRemessa(string eDir, int eNumArquivo, string eNomeArq)
        {
            var method = GetMethod<Boleto_GerarRemessa>();
            var ret = ExecuteMethod(() => method(libHandle, ToUTF8(eDir), eNumArquivo, ToUTF8(eNomeArq)));

            CheckResult(ret);
        }

        public void GerarRemessaStream(int eNumArquivo, Stream aStream)
        {
            if (aStream == null) throw new ArgumentNullException(nameof(aStream));

            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Boleto_GerarRemessaStream>();
            var ret = ExecuteMethod(() => method(libHandle, eNumArquivo, buffer, ref bufferLen));

            CheckResult(ret);

            var rem = ProcessResult(buffer, bufferLen);
            Base64ToStream(rem, aStream);
        }

        public RetornoBoleto ObterRetorno(string eDir, string eNomeArq)
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Boleto_ObterRetorno>();
            var ret = ExecuteMethod(() => method(libHandle, ToUTF8(eDir), ToUTF8(eNomeArq), buffer, ref bufferLen));

            CheckResult(ret);

            return RetornoBoleto.LerRetorno(ProcessResult(buffer, bufferLen));
        }

        public void LerRetorno(string eDir, string eNomeArq)
        {
            var method = GetMethod<Boleto_LerRetorno>();
            var ret = ExecuteMethod(() => method(libHandle, ToUTF8(eDir), ToUTF8(eNomeArq)));

            CheckResult(ret);
        }

        public string LerRetornoStream(string ARetornoBase64)
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Boleto_LerRetornoStream>();
            var ret = ExecuteMethod(() => method(libHandle, ToUTF8(ARetornoBase64), buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen);
        }

        public void EnviarEmail(string ePara, string eAssunto, string eMensagem, string eCC)
        {
            var method = GetMethod<Boleto_EnviarEmail>();
            var ret = ExecuteMethod(() => method(libHandle, ToUTF8(ePara), ToUTF8(eAssunto), ToUTF8(eMensagem), ToUTF8(eCC)));

            CheckResult(ret);
        }

        public void EnviarEmailBoleto(int eIndex, string ePara, string eAssunto, string eMensagem, string eCC)
        {
            var method = GetMethod<Boleto_EnviarEmailBoleto>();
            var ret = ExecuteMethod(() => method(libHandle, eIndex, ToUTF8(ePara), ToUTF8(eAssunto), ToUTF8(eMensagem), ToUTF8(eCC)));

            CheckResult(ret);
        }

        public void SetDiretorioArquivo(string eDir, string eArq = "")
        {
            var method = GetMethod<Boleto_SetDiretorioArquivo>();
            var ret = ExecuteMethod(() => method(libHandle, ToUTF8(eDir), ToUTF8(eArq)));

            CheckResult(ret);
        }

        public string[] ListaBancos()
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Boleto_ListaBancos>();
            var ret = ExecuteMethod(() => method(libHandle, buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen).Split('|');
        }

        public string[] ListaCaractTitulo()
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Boleto_ListaCaractTitulo>();
            var ret = ExecuteMethod(() => method(libHandle, buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen).Split('|');
        }

        public string[] ListaOcorrencias()
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Boleto_ListaOcorrencias>();
            var ret = ExecuteMethod(() => method(libHandle, buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen).Split('|');
        }

        public string[] ListaOcorrenciasEX()
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Boleto_ListaOcorrenciasEX>();
            var ret = ExecuteMethod(() => method(libHandle, buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen).Split('|');
        }

        public int TamNossoNumero(string eCarteira, string enossoNumero, string eConvenio)
        {
            var method = GetMethod<Boleto_TamNossoNumero>();
            var ret = ExecuteMethod(() => method(libHandle, ToUTF8(eCarteira), ToUTF8(enossoNumero), ToUTF8(eConvenio)));

            CheckResult(ret);

            return ret;
        }

        public string CodigosMoraAceitos()
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Boleto_CodigosMoraAceitos>();
            var ret = ExecuteMethod(() => method(libHandle, buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen);
        }

        public void SelecionaBanco(string eCodBanco)
        {
            var method = GetMethod<Boleto_SelecionaBanco>();
            var ret = ExecuteMethod(() => method(libHandle, ToUTF8(eCodBanco)));

            CheckResult(ret);
        }

        public string MontarNossoNumero(int eIndex)
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Boleto_MontarNossoNumero>();
            var ret = ExecuteMethod(() => method(libHandle, eIndex, buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen);
        }

        public string RetornaLinhaDigitavel(int eIndex)
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Boleto_RetornaLinhaDigitavel>();
            var ret = ExecuteMethod(() => method(libHandle, eIndex, buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen);
        }

        public string RetornaCodigoBarras(int eIndex)
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Boleto_RetornaCodigoBarras>();
            var ret = ExecuteMethod(() => method(libHandle, eIndex, buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen);
        }

        public RetornoWeb EnviarBoleto(OperacaoBoleto opercao)
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Boleto_EnviarBoleto>();
            var ret = ExecuteMethod(() => method(libHandle, (int)opercao, buffer, ref bufferLen));

            CheckResult(ret);

            return RetornoWeb.LerRetorno(ProcessResult(buffer, bufferLen));
        }

        public string ConsultarTitulosPorPeriodo(string eArquivoIni)
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Boleto_ConsultarTitulosPorPeriodo>();
            var ret = ExecuteMethod<int>(() => method(libHandle, eArquivoIni, buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen);
        }

        #region Private Methods

        protected override void FinalizeLib()
        {
            var finalizar = GetMethod<Boleto_Finalizar>();
            var ret = ExecuteMethod(() => finalizar(libHandle));
            CheckResult(ret);
        }

        protected override string GetUltimoRetorno(int iniBufferLen = 0)
        {
            var bufferLen = iniBufferLen < 1 ? BUFFER_LEN : iniBufferLen;
            var buffer = new StringBuilder(bufferLen);
            var ultimoRetorno = GetMethod<Boleto_UltimoRetorno>();

            if (iniBufferLen < 1)
            {
                ExecuteMethod(() => ultimoRetorno(libHandle, buffer, ref bufferLen));
                if (bufferLen <= BUFFER_LEN) return FromUTF8(buffer);

                buffer.Capacity = bufferLen;
            }

            ExecuteMethod(() => ultimoRetorno(libHandle, buffer, ref bufferLen));
            return FromUTF8(buffer);
        }

        #endregion Private Methods

        #endregion Methods
    }
}

File: Projetos/ACBrLib/Demos/Java/Boleto/Imports/ACBrLibBoletoMT/src/com/acbr/boleto/ACBrBoleto.java
package com.acbr.boleto;

import com.acbr.ACBrLibBase;
import com.acbr.ACBrSessao;
import com.sun.jna.Library;
import com.sun.jna.Native;
import com.sun.jna.Platform;
import com.sun.jna.Pointer;
import com.sun.jna.ptr.IntByReference;
import com.sun.jna.ptr.PointerByReference;
import java.io.File;
import java.nio.ByteBuffer;
import java.nio.file.Paths;

public final class ACBrBoleto extends ACBrLibBase {       
    private interface ACBrBoletoLib extends Library {
        static String JNA_LIBRARY_NAME = LibraryLoader.getLibraryName();
        public final static ACBrBoletoLib INSTANCE = LibraryLoader.getInstance();       

        class LibraryLoader {
            private static String library = "";
            private static ACBrBoletoLib instance = null;

            public static String getLibraryName() {
                if (library.isEmpty()) {
                    if(Platform.isWindows()){
                        library = Platform.is64Bit() ? "ACBrBoleto64" : "ACBrBoleto32";                        
                    }else{
                        library = Platform.is64Bit() ? "acbrboleto64" : "acbrboleto32";
                    }
                    
                }
                return library;
            }           

            public static ACBrBoletoLib getInstance() {
                if (instance == null) {
                    instance = (ACBrBoletoLib) Native.synchronizedLibrary((Library) Native.load(JNA_LIBRARY_NAME, ACBrBoletoLib.class));
                }
                return instance;
            }
        }

        int Boleto_Inicializar(PointerByReference libHandler, String eArqConfig, String eChaveCrypt);

        int Boleto_Finalizar(Pointer libHandler);

        int Boleto_Nome(Pointer libHandler, ByteBuffer buffer, IntByReference bufferSize);

        int Boleto_Versao(Pointer libHandler, ByteBuffer buffer, IntByReference bufferSize);

        int Boleto_UltimoRetorno(Pointer libHandler, ByteBuffer buffer, IntByReference bufferSize);

        int Boleto_ConfigImportar(Pointer libHandler, String eArqConfig);
        
	int Boleto_ConfigExportar(Pointer libHandler, ByteBuffer buffer, IntByReference bufferSize);
        
        int Boleto_ConfigLer(Pointer libHandler, String eArqConfig);

        int Boleto_ConfigGravar(Pointer libHandler, String eArqConfig);

        int Boleto_ConfigLerValor(Pointer libHandler, String eSessao, String eChave, ByteBuffer buffer, IntByReference bufferSize);

        int Boleto_ConfigGravarValor(Pointer libHandler, String eSessao, String eChave, String valor);

        int Boleto_ConfigurarDados(Pointer libHandler, String eArquivoIni);

        int Boleto_IncluirTitulos(Pointer libHandler, String eArquivoIni, String eTpSaida);

        int Boleto_LimparLista(Pointer libHandler); 

        int Boleto_TotalTitulosLista(Pointer libHandler);

        int Boleto_Imprimir(Pointer libHandler, String eNomeImpressora);
        
        int Boleto_ImprimirBoleto(Pointer libHandler, int eIndice, String eNomeImpressora);

        int Boleto_GerarPDF(Pointer libHandler);

        int Boleto_GerarHTML(Pointer libHandler);

        int Boleto_GerarRemessa(Pointer libHandler, String eDir, int eNumArquivo, String eNomeArquivo);

        int Boleto_LerRetorno(Pointer libHandler, String eDir, String eNomeArq);
        
        int Boleto_ObterRetorno(Pointer libHandler, String eDir, String eNomeArq, ByteBuffer buffer, IntByReference bufferSize);

        int Boleto_EnviarEmail(Pointer libHandler, String ePara, String eAssunto, String eMensagem, String eCC);

        int Boleto_EnviarEmailBoleto(Pointer libHandler, int eIndice, String ePara, String eAssunto, String eMensagem, String eCC);
        
        int Boleto_SetDiretorioArquivo(Pointer libHandler, String eDir, String eArq);

        int Boleto_ListaBancos(Pointer libHandler, ByteBuffer buffer, IntByReference bufferSize);

        int Boleto_ListaCaractTitulo(Pointer libHandler, ByteBuffer buffer, IntByReference bufferSize);

        int Boleto_ListaOcorrencias(Pointer libHandler, ByteBuffer buffer, IntByReference bufferSize);

        int Boleto_ListaOcorrenciasEX(Pointer libHandler, ByteBuffer buffer, IntByReference bufferSize);

        int Boleto_TamNossoNumero(Pointer libHandler, String eCarteira, String enossoNumero, String eConvenio);

        int Boleto_CodigosMoraAceitos(Pointer libHandler, ByteBuffer buffer, IntByReference bufferSize);

        int Boleto_SelecionaBanco(Pointer libHandler, String eCodBanco);

        int Boleto_MontarNossoNumero(Pointer libHandler, int eIndice, ByteBuffer buffer, IntByReference bufferSize);

        int Boleto_RetornaLinhaDigitavel(Pointer libHandler, int eIndice, ByteBuffer buffer, IntByReference bufferSize);

        int Boleto_RetornaCodigoBarras(Pointer libHandler, int eIndice, ByteBuffer buffer, IntByReference bufferSize);
        
        int Boleto_EnviarBoleto(Pointer libHandler, int eCodigoOperacao, ByteBuffer buffer, IntByReference bufferSize);
        
        int Boleto_ConsultarTitulosPorPeriodo(Pointer libHandler, String eArquivoIni, ByteBuffer buffer, IntByReference bufferSize);
        
    }
    
    public ACBrBoleto() throws Exception {
        File iniFile = Paths.get(System.getProperty("user.dir"), "ACBrLib.ini").toFile();
        if (!iniFile.exists()) {
            iniFile.createNewFile();
        }

        PointerByReference handle = new PointerByReference();
        int ret = ACBrBoletoLib.INSTANCE.Boleto_Inicializar(handle, toUTF8(iniFile.getAbsolutePath()), toUTF8(""));
        checkResult(ret);
		setHandle(handle.getValue());
    }

    public ACBrBoleto(String eArqConfig, String eChaveCrypt) throws Exception {
        PointerByReference handle = new PointerByReference();
        int ret = ACBrBoletoLib.INSTANCE.Boleto_Inicializar(handle, toUTF8(eArqConfig), toUTF8(eChaveCrypt));
        checkResult(ret);
		setHandle(handle.getValue());
    }
    
    @Override
    protected void dispose() throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_Finalizar(getHandle());
        checkResult(ret);
    }
    
    public String nome() throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrBoletoLib.INSTANCE.Boleto_Nome(getHandle(), buffer, bufferLen);
        checkResult(ret);

        return processResult(buffer, bufferLen);
    }

    public String versao() throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrBoletoLib.INSTANCE.Boleto_Versao(getHandle(), buffer, bufferLen);
        checkResult(ret);

        return processResult(buffer, bufferLen);
    }

    public void configLer() throws Exception {
        configLer("");
    }

    public void configLer(String eArqConfig) throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_ConfigLer(getHandle(), toUTF8(eArqConfig));
        checkResult(ret);
    }

    public void configGravar() throws Exception {
        configGravar("");
    }

    public void configGravar(String eArqConfig) throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_ConfigGravar(getHandle(), toUTF8(eArqConfig));
        checkResult(ret);
    }

    @Override
    public String configLerValor(ACBrSessao eSessao, String eChave) throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrBoletoLib.INSTANCE.Boleto_ConfigLerValor(getHandle(), toUTF8(eSessao.name()), toUTF8(eChave), buffer, bufferLen);
        checkResult(ret);
        
        return processResult(buffer, bufferLen);
    }

    @Override
    public void configGravarValor(ACBrSessao eSessao, String eChave, Object value) throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_ConfigGravarValor(getHandle(), toUTF8(eSessao.name()), toUTF8(eChave), toUTF8(value.toString()));
        checkResult(ret);
    }
    
    public void ConfigImportar(String eArqConfig) throws Exception {
        
        int ret = ACBrBoletoLib.INSTANCE.Boleto_ConfigImportar(getHandle(), eArqConfig);
        checkResult(ret);
        
    }
    
    public String ConfigExportar() throws Exception {		
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrBoletoLib.INSTANCE.Boleto_ConfigExportar(getHandle(), buffer, bufferLen);
        checkResult(ret);

        return processResult(buffer, bufferLen);		
    }
    
    public void ConfigurarDados(String eArquivoIni) throws Exception {

        int ret = ACBrBoletoLib.INSTANCE.Boleto_ConfigurarDados(getHandle(), toUTF8(eArquivoIni));
        checkResult(ret);
    }
    
    public void IncluirTitulos(String eArquivoIni, String eTpSaida) throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_IncluirTitulos(getHandle(), toUTF8(eArquivoIni), toUTF8(eTpSaida));
        checkResult(ret);
    }
    
    public void LimparLista() throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_LimparLista(getHandle());
        checkResult(ret);
    }
    
    public int TotalTitulosLista() throws Exception {

        int ret = ACBrBoletoLib.INSTANCE.Boleto_TotalTitulosLista(getHandle());
        checkResult(ret);
        
        return ret;
    }
    
    public void Imprimir() throws Exception {
        Imprimir("");
    }
    
    public void Imprimir(String eNomeImpressora) throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_Imprimir(getHandle(), eNomeImpressora);
        checkResult(ret);
    }
    
    public void ImprimirBoleto(int Indice) throws Exception {
        ImprimirBoleto(Indice, "");
    }
    
    public void ImprimirBoleto(int Indice, String eNomeImpressora) throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_ImprimirBoleto(getHandle(), Indice, eNomeImpressora);
        checkResult(ret);
    }
    
    public void GerarPDF() throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_GerarPDF(getHandle());
        checkResult(ret);
    }

    public void GerarHTML() throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_GerarHTML(getHandle());
        checkResult(ret);
    }

    public void GerarRemessa(String eDir, int eNumArquivo, String eNomeArquivo) throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_GerarRemessa(getHandle(), eDir, eNumArquivo, eNomeArquivo);
        checkResult(ret);
    }
    
    public void LerRetorno(String eDir, String eNomeArq) throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_LerRetorno(getHandle(), eDir, eNomeArq);
        checkResult(ret);
    }
    
    public String ObterRetorno(String eDir, String eNomeArq) throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);
        
        int ret = ACBrBoletoLib.INSTANCE.Boleto_ObterRetorno(getHandle(), eDir, eNomeArq, buffer, bufferLen);
        checkResult(ret);
        
        return processResult(buffer, bufferLen);
    }

    public void EnviarEmail(String ePara, String eAssunto, String eMensagem, String eCC) throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_EnviarEmail(getHandle(), ePara, eAssunto, eMensagem, eCC);
        checkResult(ret);
    }
    
    public void EnviarEmailBoleto(int eIndice, String ePara, String eAssunto, String eMensagem, String eCC ) throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_EnviarEmailBoleto(getHandle(),eIndice, ePara, eAssunto, eMensagem, eCC);
        checkResult(ret);
    }
    
    public void SetDiretorioArquivo(String eDir, String eArq) throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_SetDiretorioArquivo(getHandle(), eDir, eArq);
        checkResult(ret);
    }

    public String[] ListaBancos() throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrBoletoLib.INSTANCE.Boleto_ListaBancos(getHandle(), buffer, bufferLen);
        checkResult(ret);
        
        return processResult(buffer, bufferLen).split("|");
    }
    
    public String[] ListaCaractTitulo() throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrBoletoLib.INSTANCE.Boleto_ListaCaractTitulo(getHandle(), buffer, bufferLen);
        checkResult(ret);
        
        return processResult(buffer, bufferLen).split("|");
    }
    
    public String ListaOcorrencias() throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrBoletoLib.INSTANCE.Boleto_ListaOcorrencias(getHandle(), buffer, bufferLen);
        checkResult(ret);
        
        return processResult(buffer, bufferLen);
    }

    public String[] ListaOcorrenciasEX() throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrBoletoLib.INSTANCE.Boleto_ListaOcorrenciasEX(getHandle(), buffer, bufferLen);
        checkResult(ret);
        
        return processResult(buffer, bufferLen).split("|");
    }
    
    public int TamNossoNumero(String eCarteira, String enossoNumero, String eConvenio) throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_TamNossoNumero(getHandle(), eCarteira, enossoNumero, eConvenio);
        checkResult(ret);
        
        return ret;
    }
    
    public String CodigosMoraAceitos() throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrBoletoLib.INSTANCE.Boleto_CodigosMoraAceitos(getHandle(), buffer, bufferLen);
        checkResult(ret);
        
        return processResult(buffer, bufferLen);
    }
    
    public void SelecionaBanco(String eCodBanco) throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_SelecionaBanco(getHandle(), eCodBanco);
        checkResult(ret);
    }

    public String MontarNossoNumero(int eIndice) throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrBoletoLib.INSTANCE.Boleto_MontarNossoNumero(getHandle(), eIndice, buffer, bufferLen);
        checkResult(ret);
        
        return processResult(buffer, bufferLen);
    }

    public String RetornaLinhaDigitavel(int eIndice) throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrBoletoLib.INSTANCE.Boleto_RetornaLinhaDigitavel(getHandle(), eIndice, buffer, bufferLen);
        checkResult(ret);
        
        return processResult(buffer, bufferLen);
    }

    public String RetornaCodigoBarras(int eIndice) throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrBoletoLib.INSTANCE.Boleto_RetornaCodigoBarras(getHandle(), eIndice, buffer, bufferLen);
        checkResult(ret);
        
        return processResult(buffer, bufferLen);
    }    
    
    public String EnviarBoleto(int eCodigoOperacao) throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);
        
        int ret = ACBrBoletoLib.INSTANCE.Boleto_EnviarBoleto(getHandle(), eCodigoOperacao, buffer, bufferLen);
        checkResult(ret);
        
        return processResult(buffer, bufferLen);
    }
    
    public String ConsultarTitulosPorPeriodo(String eArquivoIni) throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);
        
        int ret = ACBrBoletoLib.INSTANCE.Boleto_ConsultarTitulosPorPeriodo(getHandle(), eArquivoIni, buffer, bufferLen);
        checkResult(ret);
        
        return processResult(buffer, bufferLen);
    }
    
    @Override
    protected void UltimoRetorno(ByteBuffer buffer, IntByReference bufferLen) {
        ACBrBoletoLib.INSTANCE.Boleto_UltimoRetorno(getHandle(), buffer, bufferLen);
    }
}
